src/Netscript/APIWrapper.ts

```diff
@@ -0,0 +1,157 @@
+  string: (argName: string, v: unknown) => string;
+  number: (argName: string, v: unknown) => number;
+  boolean: (v: unknown) => boolean;
+  getServer: (hostname: string) => BaseServer;
+  checkSingularityAccess: () => void;
+  hack: (hostname: any, manual: any, { threads: requestedThreads, stock }?: any) => Promise<number>;
+  getValidPort: (port: any) => IPort;
+};
+
+function wrapFunction(
+  helpers: NetscriptHelpers,
+  wrappedAPI: any,
+  workerScript: WorkerScript,
+  func: (_ctx: NetscriptContext) => (...args: unknown[]) => unknown,
+  ...tree: string[]
+): void {
+  const functionPath = tree.join(".");
+  const functionName = tree.pop();
+  if (typeof functionName !== "string") {
+    throw makeRuntimeRejectMsg(workerScript, "Failure occured while wrapping netscript api");
+  }
+  const ctx = {
+    makeRuntimeErrorMsg: (message: string) => {
+      return helpers.makeRuntimeErrorMsg(functionPath, message);
+    },
+    log: (message: () => string) => {
+      workerScript.log(functionPath, message);
+    },
+    workerScript,
+    function: functionName,
+    helper: {
+      updateDynamicRam: (ramCost: number) => helpers.updateDynamicRam(functionName, ramCost),
+      makeRuntimeErrorMsg: (msg: string) => helpers.makeRuntimeErrorMsg(functionPath, msg),
+      string: (argName: string, v: unknown) => helpers.string(functionPath, argName, v),
+      number: (argName: string, v: unknown) => helpers.number(functionPath, argName, v),
+      boolean: helpers.boolean,
+      getServer: (hostname: string) => helpers.getServer(hostname, functionPath),
+      checkSingularityAccess: () => helpers.checkSingularityAccess(functionName),
+      hack: helpers.hack,
+      getValidPort: (port: any) => helpers.getValidPort(functionPath, port),
+    },
+  };
+  function wrappedFunction(...args: unknown[]): unknown {
+    helpers.updateDynamicRam(ctx.function, getRamCost(Player, ...tree, ctx.function));
+    return func(ctx)(...args);
+  }
+  const parent = getNestedProperty(wrappedAPI, ...tree);
+  Object.defineProperty(parent, functionName, {
+    value: wrappedFunction,
+    writable: true,
+    enumerable: true,
+  });
+}
+
+export function wrapAPI(
+  helpers: NetscriptHelpers,
+  wrappedAPI: ExternalAPI,
+  workerScript: WorkerScript,
+  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+  namespace: any,
+  ...tree: string[]
+): WrappedNetscriptAPI {
+  if (typeof namespace !== "object") throw new Error("Invalid namespace?");
+  for (const property of Object.getOwnPropertyNames(namespace)) {
+    switch (typeof namespace[property]) {
+      case "function": {
+        wrapFunction(helpers, wrappedAPI, workerScript, namespace[property], ...tree, property);
+        break;
+      }
+      case "object": {
+        wrapAPI(helpers, wrappedAPI, workerScript, namespace[property], ...tree, property);
+        break;
+      }
+      default: {
+        setNestedProperty(wrappedAPI, namespace[property], ...tree, property);
```

Comment:
This sends 4 arguments but the function declares 3 arguments

Reply:
last arguent in setNestedProperty is a rest argument, so it takes all of tree and property
ie tree = ['a','b','c']; property = 'd';
inside the setNestedProperty call the tree argument contains ['a','b','c','d']